src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,157 @@
+#include "greedy_selector.hpp"
+
+#include <algorithm>
+#include <limits>
+#include <list>
+
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+/*
+ * This subroutine scans the least desirable available choices until the
+ * accumulated reject_cost() is lower than a required value.
+ * If the thereby accumulated reject_desirability is not lower than a threshold
+ * an iterator to the choice one past the subsequence to reject is returned.
+ * Otherwise the end() iterator is returned, that the constraints are not satisfied
+ * by any continuous subsequence starting at sorted_choices.begin().
+ */
+const std::list<std::shared_ptr<TuningChoice>>::const_iterator sacrifice_choices(
+    const std::list<std::shared_ptr<TuningChoice>>& sorted_choices, float required_cost_delta,
+    float acceptible_desirability_delta = -std::numeric_limits<float>::infinity()) {
+  auto desirability_delta = 0.0f;
+  auto cost_delta = 0.0f;
+
+  auto choice = sorted_choices.cbegin();
+  while (choice != sorted_choices.cend()) {
+    if (cost_delta <= required_cost_delta) {
+      break;
+    }
+    desirability_delta += (*choice)->reject_desirability();
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.cend();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
```

Comment:
Instead of writing a comment, you could use DebugAssert to actually check if your assumptions hold true

Reply:
oh yeah

Reply:
Added the necessary asserts in the balance accumulation loop at the beginning of this method.